drivers/ethernet/eth_mcux.c

```diff
@@ -647,7 +651,7 @@ static bool eth_get_ptp_data(struct net_if *iface, struct net_pkt *pkt)
 }
 #endif /* CONFIG_PTP_CLOCK_MCUX */
 
-static int eth_tx(const struct device *dev, struct net_pkt *pkt)
+int eth_tx(const struct device *dev, struct net_pkt *pkt)
```

Comment:
We should keep this as static, driver functions are suppose to be private.

Reply:
If I remember correctly - the last recommendation was to introduce something like:
`struct dsa_api` structure with `.send` member.

Then I would instantiate this structure in eth_mcux.c only when CONFIG_NET_DSA is defined and set
`send = eth_tx`

In that way eth_tx could stay `static` (private) in eth_mcux.c

Reply:
> If I remember correctly - the last recommendation was to introduce something like:
> `struct dsa_api` structure with `.send` member.

That sounds good idea.

Reply:
**Another option would be to extend `struct ethernet_api`** from `./include/net/ethernet.h` to have:
`#if defined(CONFIG_NET_DSA)`
`	/** Send a network packet via DSA master port */`
`	int (*dsa_master_send)(const struct device *dev, struct net_pkt *pkt);`
`#endif`

The dsa_master_send callback could be set in `eth_mcux.c` and reused in DSA. From other comments you also proposed to add some MAC address related functions into the same structure.

**Yet another option would be to extend `struct ethernet_context`** in the same file to have
`/** Master ETH port's transmit callback - necessary for sending`
`* packets with DSA switch tag appended */`
`eth_tx_t master_eth_tx;`

Where: `typedef int (*eth_tx_t) (struct device *dev, struct net_pkt *pkt);`
and `eth_tx_t` corresponds to `eth_tx` from `eth_mcux.c`

**The question:**
I'm puzzled which place would be better. The `eth_ctx` is easier accessible from iface, but is harder to setup in `eth_mcux.c` (i.e. ETH driver files).

Any preferences?

Reply:
After some thoughts - IMHO it would be better to extend the struct ethernet_api as we will add some other callbacks (which e.g. set MAC address) there as well.

Reply:
I was thinking more of a 3rd option:

Create `dsa_api` which extends `ethernet_api`. Then in `dsa_ksz8794.c`, instead of using `ethernet_api`, you would set
```
const struct dsa_api dsa_api_funcs = {
	.ethernet_api.iface_api.init	= dsa_iface_init,
	.ethernet_api.get_capabilities	= dsa_port_get_capabilities,
	.ethernet_api.send		= dsa_tx,
// add here all the dsa specific APIs you need
};
```
This way the driver would still be Ethernet compatible as it would have all the Ethernet callbacks in it, but we would avoid adding dsa specific stuff into ethernet_api struct.
Does this sound feasible?

Reply:
I've spent some time to provide implementation for this idea.
Unfortunately, it causes circular data dependencies between `./include/net/dsa.h` and `ethernet.h`.
The `struct ethernet_context` is using dsa callback exported in `./include/net/dsa.h` and `struct dsa_api` (from `dfsa.h`) uses `struct ethernet_api eapi;` defined in `ethernet.h`.

As fair as I can observe - the `ethernet.h` uses (and includes) data from other headers (like `net/lldp.h, net/ethernet_vlan.h`). In that way `struct ethernet_api` has callbacks for VLAN and PTP. 

What works with very little adjustments is to augment the `struct ethernet_api` with:
`#if defined(CONFIG_NET_DSA)`
`	/** DSA specific API callbacks */`
`	struct dsa_api dapi;`
`#endif`

And put all DSA related callbacks (also HW specific) into `struct dsa_api dapi`.

Reply:
It there are circular dependencies, we would need to fix it then. But I know these can be tricky to solve.
Anyway, please propose then embedding dsa to ethernet_api, lets see how it looks like.

Reply:
Please find updated PR